code/datums/chatmessage.dm

```diff
@@ -0,0 +1,87 @@
+#define CHAT_MESSAGE_SPAWN_TIME		1  // 0.1 second
+#define CHAT_MESSAGE_LIFESPAN		50 // 5 seconds
+#define CHAT_MESSAGE_EOL_FADE		10 // 1 second
+#define CHAT_MESSAGE_WIDTH			96 // pixels
+#define CHAT_MESSAGE_MAX_LENGTH		110 // characters
+#define WXH_TO_HEIGHT(x)			text2num(copytext((x), findtextEx((x), "x") + 1)) // thanks lummox
+
+/client
+	/// Messages currently seen by this client
+	var/list/seen_messages = list()
+
+/**
+  * # Chat Message Overlay
+  *
+  * Datum for generating a message overlay on the map
+  */
+/datum/chatmessage
+	/// The visual element of the chat messsage
+	var/image/message
+	/// The location in which the message is appearing
+	var/atom/message_loc
+	/// The client who heard this message
+	var/client/owned_by
+
+/**
+  * Constructs a chat message overlay
+  *
+  * Arguments:
+  * * text - The text content of the overlay
+  * * target - The target atom to display the overlay at
+  * * owner - The mob that owns this overlay, only this mob will be able to view it
+  * * extra_classes - Extra classes to apply to the span that holds the text
+  * * lifespan - The lifespan of the message in deciseconds
+  */
+/datum/chatmessage/New(text, atom/target, mob/owner, list/extra_classes = null, lifespan = CHAT_MESSAGE_LIFESPAN)
+	. = ..()
+	if (!istype(target))
+		EXCEPTION("Invalid target given for chatmessage")
+	if(QDELETED(owner) || !istype(owner) || !owner.client)
+		stack_trace("/datum/chatmessage created with [isnull(owner) ? "null" : "invalid"] mob owner")
+		qdel(src)
+		return
+
+	// Clip message
+	if (length(text) > CHAT_MESSAGE_MAX_LENGTH)
+		text = copytext(text, 1, CHAT_MESSAGE_MAX_LENGTH) + "..."
```

Comment:
These are byte counts, not char counts. They should be length_char() and copytext_char().